### PR TITLE
[FLPTH-156] Adds a workflow to migrate a VM from vSphere to OCP Virtualization using Forklift

### PR DESCRIPTION
**Pull request by jordigilh**
Includes an integration test using the SDK. However, in order to run the test locally you'll need to have access to an OpenShift cluster with OCP Virtualization, Forklift and access to a vSphere environment with a VM that can be migrated. 

The java class in the integration test is not run as part of the CI/CD because of these dependencies.

### CODE REVIEW

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/checker/MigrationStatusWorkFlowChecker.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+			return new DefaultWorkReport(WorkStatus.FAILED, workContext, e);
+		}
+		catch (MissingParameterException e) {
+			return new DefaultWorkReport(WorkStatus.FAILED, workContext, e);
+		}
+
+		return new DefaultWorkReport(WorkStatus.FAILED, workContext);
+	}
+
+	public List<WorkParameter> getWorkFlowTaskParameters() {
```
Comment (gciavarrini) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/checker/MigrationStatusWorkFlowChecker.java:71]: If I'm not wrong the proper definition of this function should be:

@Override
public @NonNull List<WorkParameter> getWorkFlowTaskParameters() {
....
}

can you please double check?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/assessment/VmMigrationWorkFlowTask.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+			String vmName = getRequiredParameterValue(Constants.VM_NAME_PARAMETER_NAME);
+			V1beta1Plan plan = Kubernetes.createPlanForVM(vmName); // mtv-rhel8-sanity
+			CustomObjectsApi coapi = new CustomObjectsApi(client);
+			coapi.createNamespacedCustomObject("forklift.konveyor.io", "v1beta1", "demo24", "plans", plan, null, null,
+					null);
+			// Object resp = coapi.listNamespacedCustomObject("forklift.konveyor.io",
+			// "v1beta1", "demo24",
+			// "plans",null,false,null,null,null,null,null,null,null,false);
+			// Store plan's name in the context
+			addParameter(Constants.PLAN_NAME_PARAMETER_NAME, plan.getMetadata().getName());
```
Comment (gciavarrini) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/assessment/VmMigrationWorkFlowTask.java:56]: `getMetadata` is nullable

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/assessment/VmMigrationWorkFlowTask.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			addParameter(Constants.KUBERNETES_API_SERVER_URL, apiURL);
+			addParameter(Constants.KUBERNETES_TOKEN, token);
+			if (caCert != null) {
+				addParameter(Constants.KUBERNETES_CA_CERT, caCert);
+			}
+		}
+		catch (ApiException e) {
+			log.error("Failed to create plan", e);
+			ApiException a = (ApiException) e;
+			log.error(a.getResponseBody());
```
Comment (gciavarrini) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/assessment/VmMigrationWorkFlowTask.java:66]: debug leftover?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/assessment/VmMigrationWorkFlowTask.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		}
+		catch (MissingParameterException e) {
+			log.error("Can't get parameter {} value", Constants.VM_NAME_PARAMETER_NAME);
+			return new DefaultWorkReport(WorkStatus.FAILED, workContext);
+		}
+
+		return new DefaultWorkReport(WorkStatus.COMPLETED, workContext);
+	}
+
+	public List<WorkParameter> getWorkFlowTaskParameters() {
```
Comment (gciavarrini) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/assessment/VmMigrationWorkFlowTask.java:77]: If I'm not wrong the proper definition of this function should be:

@OverRide
public @nonnull List getWorkFlowTaskParameters() {
....
}

can you please double check?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/VmMigrationWorkFlowConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	// next. In this case there is only one.
+	@Bean(name = "vmMigration" + WorkFlowConstants.ASSESSMENT_WORKFLOW)
+	@Assessment(parameters = {
+			@Parameter(key = Constants.VM_NAME_PARAMETER_NAME,
+					description = "The name of the VM to migrate from vmware to OCP Virtualization",
+					type = WorkParameterType.TEXT, optional = false),
+			@Parameter(key = Constants.KUBERNETES_API_SERVER_URL, description = "The URL of the kubernetes API server",
+					type = WorkParameterType.TEXT, optional = false),
+			@Parameter(key = Constants.KUBERNETES_TOKEN,
+					description = "The token to authenitcate against the kubernetes API server",
```
Comment (gciavarrini) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/VmMigrationWorkFlowConfiguration.java:60]: s/authenitcate/authenticate/

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/VmMigrationWorkFlowConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+	@Bean(name = "vmMigrationWorkFlow" + WorkFlowConstants.INFRASTRUCTURE_WORKFLOW)
+	@Infrastructure(parameters = {
+			@Parameter(key = Constants.VM_NAME_PARAMETER_NAME,
+					description = "The name of the VM to migrate from vmware to OCP Virtualization",
+					type = WorkParameterType.TEXT, optional = false),
+			@Parameter(key = Constants.KUBERNETES_API_SERVER_URL, description = "The URL of the kubernetes API server",
+					type = WorkParameterType.TEXT, optional = false),
+			@Parameter(key = Constants.KUBERNETES_TOKEN,
+					description = "The token to authenitcate against the kubernetes API server",
```
Comment (gciavarrini) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/VmMigrationWorkFlowConfiguration.java:127]: s/authenitcate/authenticate/

File: integration-tests/src/test/java/com/redhat/parodos/flows/VMMigrationWorkFlow.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+		log.info("The Following Option Is Available: {}", infrastructureOption);
+
+		workFlowRequestDTO = new WorkFlowRequestDTO();
+		workFlowRequestDTO.setProjectId(components.project().getId());
+		workFlowRequestDTO.setWorkFlowName(infrastructureOption);
+
+		workFlowRequestDTO.addArgumentsItem(vmName).addArgumentsItem(apiUrl).addArgumentsItem(token);
+
+		log.info("******** Start VM Migration task  ********");
+		workFlowResponseDTO = execute(workflowApi, workFlowRequestDTO);
```
Comment (gciavarrini) [inline integration-tests/src/test/java/com/redhat/parodos/flows/VMMigrationWorkFlow.java:72]: WDYT to assert this object?

In addition, our code style prefer `assertThat(xyz).isNotNull()` style. Can you please update the assertions?

**Review by gciavarrini — COMMENTED** [review]
Can you please add a section in `README.md`?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/dto/models/V1beta1Migration.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
+/*
+ * Kubernetes
+ * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
+ *
+ * The version of the OpenAPI document: v1.21.1
```
Comment (masayag) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/dto/models/V1beta1Migration.java:5]: How was this class generated? should we include the openapi + generating this file as part of our code base or is there an option to rely on external lib for this?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/task/CreateMigrationWorkFlowTask.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+			}
+		}
+		catch (ApiException e) {
+			log.error("Failed to create plan or migration", e);
+			ApiException a = (ApiException) e;
+			log.error(a.getResponseBody());
+			return new DefaultWorkReport(WorkStatus.FAILED, workContext, e);
+		}
+		catch (MissingParameterException e) {
+			log.error("Can't get parameter {} value", Constants.VM_NAME_PARAMETER_NAME);
```
Comment (masayag) [inline workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/task/CreateMigrationWorkFlowTask.java:64]: the message will be accurate if the following fails:
```
getRequiredParameterValue(Constants.KUBERNETES_TOKEN);
```
I believe `MissingParameterException` should be extended to record also the missed variable name and produce a standard message for all errors to be used, but not in the scope for this PR.

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/utils/Kubernetes.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+		return migration;
+	}
+
+	public static V1beta1Plan createPlanForVM(String vmName) {
+		V1beta1PlanSpecMapStorage storage = new V1beta1PlanSpecMapStorage();
+		storage.setName("vmware-qfs8w");
+		storage.namespace("demo24");
+		V1beta1PlanSpecMapNetwork network = new V1beta1PlanSpecMapNetwork();
+		network.setName("vmware-n9lxw");
+		network.setNamespace("demo24");
```
[the remark this answers is not in the document]
Comment (masayag) [reply workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/utils/Kubernetes.java:67]: +1, if this class acts as a utility class, it should expose the parameters to be provided by the workflow

**Review by masayag — CHANGES_REQUESTED** [review]

**Comment by masayag** [comment]
General question: why were the workflow and tasks added under workflow-examples and not as a prebuilt-task?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/dto/models/V1beta1Migration.java (hunk 8 above)
Comment (jordigilh) [reply workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/dto/models/V1beta1Migration.java]: Eventually I migrated to use fabric8 to generate the objects, so I had to recreate them from scratch.

**Comment by jordigilh** [comment]
> Can you please add a section in `README.md`?

@gciavarrini what would you like me to add in the `workflow-examples/README.md`? I'm not sure what to add.

**Comment by jordigilh** [comment]
> General question: why were the workflow and tasks added under workflow-examples and not as a prebuilt-task?

I started working directly on the workflow-examples and I was not aware of the differences to move it later on. Do you think I should move it there?

**Comment by gciavarrini** [comment]
@jordigilh 
My suggestion is to document a bit this example to let users quickly understand what you did. Maybe we can add page "Workflow examples" on the website. 
I't's a way to share what know/learn during this PR creation among the team members and the users.
WDYT?

**Comment by jordigilh** [comment]
> @jordigilh My suggestion is to document a bit this example to let users quickly understand what you did. Maybe we can add page "Workflow examples" on the website. I't's a way to share what know/learn during this PR creation among the team members and the users. WDYT?

I can add a javadoc to the Configuration class that can be used later on in a readme or website. I think your suggestion to have it either on the readme or the website is a good one, just not sure it should be applied only to this code on this PR, but rather to have all examples documented equally and added at the same time in a single PR.

**Comment by gciavarrini** [comment]
> I started working directly on the workflow-examples and I was not aware of the differences to move it later on. Do you think I should move it there?

In such case, we already documented other prebuilt tasks https://www.parodos.dev/documentation/latest/operations/prebuilt.html :)

**Comment by jordigilh** [comment]
> > I started working directly on the workflow-examples and I was not aware of the differences to move it later on. Do you think I should move it there?
> 
> In such case, we already documented other prebuilt tasks https://www.parodos.dev/documentation/latest/operations/prebuilt.html :)

I'll move it to prebuilt-tasks and update the doc :smile:

**Comment by pkliczewski** [comment]
@masayag is there anything else you would like to see updated?

File: workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/VmMigrationWorkFlowConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+					type = WorkParameterType.TEXT, optional = false),
+			@Parameter(key = Constants.KUBERNETES_CA_CERT,
+					description = "The Certificate Authority Cert to validate the encrypted connection with the API server",
```
[the remark this answers is not in the document]
Comment (masayag) [reply workflow-examples/src/main/java/com/redhat/parodos/examples/vmmigration/VmMigrationWorkFlowConfiguration.java]: perhaps base64 format

**Pull request merged**